### PR TITLE
[ISSUE #3949] Support gRPC protocol

### PR DESCRIPTION
**Pull request by drpmma**
## What is the purpose of the change

Support gRPC protocol in rocketmq with `Local` mode and `Cluster` mode.

## Brief changelog

* Support gRPC protocol.
* Add new module `rocketmq-proxy`
* Implement two gRPC protocol implementations, `Local` mode and `Cluster` mode.
* Support ACL with gRPC protocol
* Fix bugs about `POP_MESSAGE`
* Add `ChangeListener` for `ProducerManager` and `ConsumerManager`
* Add `toString` method for RequestHeader

### CODE REVIEW

**Comment by lollipopjin** [comment]
For details, see [[RIP-39] Support gRPC protocol](https://github.com/apache/rocketmq/issues/3949)

**Comment by hill007299** [comment]
hi  @drpmma  please check and fix the build failure ，https://github.com/apache/rocketmq/pull/4225/checks?check_run_id=6226106913

**Comment by drpmma** [comment]
> 

It require the release of org.apache.rocketmq:rocketmq-proto:jar:2.0.0-SNAPSHOT.

File: acl/src/main/java/org/apache/rocketmq/acl/plain/PlainAccessResource.java [hunk 1]
```diff
@@ -55,6 +86,187 @@ public class PlainAccessResource implements AccessResource {
...
+            return accessResource;
+        }
+        accessResource.setAccessKey(request.getExtFields().get(SessionCredentials.ACCESS_KEY));
+        accessResource.setSignature(request.getExtFields().get(SessionCredentials.SIGNATURE));
+        accessResource.setSecretToken(request.getExtFields().get(SessionCredentials.SECURITY_TOKEN));
+
+        try {
+            switch (request.getCode()) {
+                case RequestCode.SEND_MESSAGE:
+                    accessResource.addResourceAndPerm(request.getExtFields().get("topic"), Permission.PUB);
```
Comment (hill007299) [inline acl/src/main/java/org/apache/rocketmq/acl/plain/PlainAccessResource.java:111]: Please use string constants

File: acl/src/main/java/org/apache/rocketmq/acl/common/AuthenticationHeader.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    @Override public String toString() {
+        final StringBuilder sb = new StringBuilder("MetadataHeader{");
+        sb.append("remoteAddress='").append(remoteAddress).append('\'');
+        sb.append(", tenantId='").append(tenantId).append('\'');
+        sb.append(", namespace='").append(namespace).append('\'');
+        sb.append(", authorization='").append(authorization).append('\'');
+        sb.append(", datetime='").append(datetime).append('\'');
+        sb.append(", sessionToken='").append(sessionToken).append('\'');
+        sb.append(", requestId='").append(requestId).append('\'');
```
Comment (zhouxinyu) [inline acl/src/main/java/org/apache/rocketmq/acl/common/AuthenticationHeader.java:227]: Use a unified ToStringBuilder.

Comment (xdkxlk) [reply acl/src/main/java/org/apache/rocketmq/acl/common/AuthenticationHeader.java]: done

**Review by zhouxinyu — APPROVED** [review]
Let's merge this PR.

**Review by lollipopjin — APPROVED** [review]
LGTM

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java [hunk 3]
```diff
@@ -210,8 +210,8 @@ public void initChannel(SocketChannel ch) throws Exception {
                             LOGGER.warn("Connections are insecure as SSLContext is null!");
                         }
                     }
-                    pipeline.addLast(
-                        defaultEventExecutorGroup,
```
Comment (lizhanhui) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java:214]: If null is passed in, the follow-up handlers would be skipped. This is a bug. 

See https://github.com/netty/netty/blob/d34212439068091bcec29a8fad4df82f0a82c638/transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java#L377

**Comment by lizhanhui** [comment]
It's a really large PR... Let's merge it as an evolution starting point.

**Pull request merged**